File: generated_tasks/nexusflow__77dc75b/environment/repo/nexusflow/telemetry/metrics.py
"""
nexusflow.telemetry.metrics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Metrics collection framework with counters, gauges, histograms,
and timers. Supports labels, aggregation, and metric export.
"""

from __future__ import annotations

import math
import threading
import time
from collections import defaultdict
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable, Dict, List, Optional, Set, Tuple


class MetricType(Enum):
    COUNTER = "counter"
    GAUGE = "gauge"
    HISTOGRAM = "histogram"
    TIMER = "timer"


@dataclass
class MetricLabel:
    """A set of key-value labels for a metric."""
    labels: Dict[str, str] = field(default_factory=dict)

    @property
    def key(self) -> str:
        sorted_items = sorted(self.labels.items())
        return ",".join(f"{k}={v}" for k, v in sorted_items)


class Counter:
    """A monotonically increasing counter."""

    def __init__(self, name: str, description: str = "") -> None:
        self.name = name
        self.description = description
        self.metric_type = MetricType.COUNTER
        self._values: Dict[str, float] = defaultdict(float)
        self._lock = threading.Lock()

    def inc(self, amount: float = 1.0, labels: Optional[Dict[str, str]] = None) -> None:
        """Increment the counter."""
        if amount < 0:
            raise ValueError("Counter can only be incremented")
        key = MetricLabel(labels or {}).key
        with self._lock:
            self._values[key] += amount

    def get(self, labels: Optional[Dict[str, str]] = None) -> float:
        key = MetricLabel(labels or {}).key
        return self._values.get(key, 0.0)

    def reset(self) -> None:
        with self._lock:
            self._values.clear()

    def collect(self) -> Dict[str, float]:
        with self._lock:
            return dict(self._values)


class Gauge:
    """A value that can go up and down."""

    def __init__(self, name: str, description: str = "") -> None:
        self.name = name
        self.description = description
        self.metric_type = MetricType.GAUGE
        self._values: Dict[str, float] = defaultdict(float)
        self._lock = threading.Lock()

    def set(self, value: float, labels: Optional[Dict[str, str]] = None) -> None:
        key = MetricLabel(labels or {}).key
        with self._lock:
            self._values[key] = value

    def inc(self, amount: float = 1.0, labels: Optional[Dict[str, str]] = None) -> None:
        key = MetricLabel(labels or {}).key
        with self._lock:
            self._values[key] += amount

    def dec(self, amount: float = 1.0, labels: Optional[Dict[str, str]] = None) -> None:
        key = MetricLabel(labels or {}).key
        with self._lock:
            self._values[key] -= amount

    def get(self, labels: Optional[Dict[str, str]] = None) -> float:
        key = MetricLabel(labels or {}).key
        return self._values.get(key, 0.0)

    def collect(self) -> Dict[str, float]:
        with self._lock:
            return dict(self._values)


class Histogram:
    """Records observations in configurable buckets."""

    DEFAULT_BUCKETS = (0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0)

    def __init__(
        self,
        name: str,
        description: str = "",
        buckets: Optional[Tuple[float, ...]] = None,
    ) -> None:
        self.name = name
        self.description = description
        self.metric_type = MetricType.HISTOGRAM
        self._buckets = sorted(buckets or self.DEFAULT_BUCKETS)
        self._observations: Dict[str, List[float]] = defaultdict(list)
        self._bucket_counts: Dict[str, Dict[float, int]] = {}
        self._lock = threading.Lock()

    def observe(self, value: float, labels: Optional[Dict[str, str]] = None) -> None:
        """Record an observation."""
        key = MetricLabel(labels or {}).key
        with self._lock:
            self._observations[key].append(value)
            if key not in self._bucket_counts:
                self._bucket_counts[key] = {b: 0 for b in self._buckets}
            for bucket in self._buckets:
                if value <= bucket:
                    self._bucket_counts[key][bucket] += 1

    def get_count(self, labels: Optional[Dict[str, str]] = None) -> int:
        key = MetricLabel(labels or {}).key
        return len(self._observations.get(key, []))

    def get_sum(self, labels: Optional[Dict[str, str]] = None) -> float:
        key = MetricLabel(labels or {}).key
        return sum(self._observations.get(key, []))

    def get_mean(self, labels: Optional[Dict[str, str]] = None) -> float:
        key = MetricLabel(labels or {}).key
        obs = self._observations.get(key, [])
        if not obs:
            return 0.0
        return sum(obs) / len(obs)

    def get_percentile(
        self, percentile: float, labels: Optional[Dict[str, str]] = None
    ) -> float:
        """Calculate a percentile from observations."""
        key = MetricLabel(labels or {}).key
        obs = sorted(self._observations.get(key, []))
        if not obs:
            return 0.0
        idx = int(len(obs) * percentile / 100.0)
        idx = min(idx, len(obs) - 1)
        return obs[idx]

    def get_buckets(self, labels: Optional[Dict[str, str]] = None) -> Dict[str, int]:
        key = MetricLabel(labels or {}).key
        counts = self._bucket_counts.get(key, {})
        return {f"le_{b}": c for b, c in sorted(counts.items())}

    def collect(self) -> Dict[str, Any]:
        with self._lock:
            result: Dict[str, Any] = {}
            for key in self._observations:
                obs = self._observations[key]
                result[key] = {
                    "count": len(obs),
                    "sum": sum(obs),
                    "mean": sum(obs) / len(obs) if obs else 0,
                    "buckets": self.get_buckets(None),
                }
            return result


class Timer:
    """Measures duration of operations."""

    def __init__(self, name: str, description: str = "") -> None:
        self.name = name
        self.description = description
        self._histogram = Histogram(f"{name}_seconds", description)

    def time(self, labels: Optional[Dict[str, str]] = None) -> "TimerContext":
        """Context manager to time an operation."""
        return TimerContext(self._histogram, labels)

    def observe(self, duration: float, labels: Optional[Dict[str, str]] = None) -> None:
        self._histogram.observe(duration, labels)

    def collect(self) -> Dict[str, Any]:
        return self._histogram.collect()


class TimerContext:
    """Context manager for timing operations."""

    def __init__(self, histogram: Histogram, labels: Optional[Dict[str, str]] = None) -> None:
        self._histogram = histogram
        self._labels = labels
        self._start: float = 0.0

    def __enter__(self) -> "TimerContext":
        self._start = time.time()
        return self

    def __exit__(self, *args: Any) -> None:
        duration = time.time() - self._start
        self._histogram.observe(duration, self._labels)


class MetricsRegistry:
    """Central registry for all metrics."""

    def __init__(self) -> None:
        self._metrics: Dict[str, Any] = {}
        self._lock = threading.Lock()

    def counter(self, name: str, description: str = "") -> Counter:
        with self._lock:
            if name not in self._metrics:
                self._metrics[name] = Counter(name, description)
            return self._metrics[name]

    def gauge(self, name: str, description: str = "") -> Gauge:
        with self._lock:
            if name not in self._metrics:
                self._metrics[name] = Gauge(name, description)
            return self._metrics[name]

    def histogram(
        self, name: str, description: str = "",
        buckets: Optional[Tuple[float, ...]] = None,
    ) -> Histogram:
        with self._lock:
            if name not in self._metrics:
                self._metrics[name] = Histogram(name, description, buckets)
            return self._metrics[name]

    def timer(self, name: str, description: str = "") -> Timer:
        with self._lock:
            if name not in self._metrics:
                self._metrics[name] = Timer(name, description)
            return self._metrics[name]

    def collect_all(self) -> Dict[str, Any]:
        """Collect all metrics."""
        result: Dict[str, Any] = {}
        for name, metric in self._metrics.items():
            result[name] = {
                "type": metric.metric_type.value if hasattr(metric, "metric_type") else "unknown",
                "values": metric.collect(),
            }
        return result

    def get_metric_names(self) -> List[str]:
        return list(self._metrics.keys())

    def clear(self) -> None:
        with self._lock:
            self._metrics.clear()

File: generated_tasks/swarmsync__crdt-merge-convergence/environment/repo/pkg/metrics/metrics.go
package metrics

import (
	"math"
	"sort"
	"sync"
	"sync/atomic"
)

// Counter is a monotonically increasing counter.
type Counter struct {
	value atomic.Int64
	name  string
}

func NewCounter(name string) *Counter { return &Counter{name: name} }
func (c *Counter) Inc()               { c.value.Add(1) }
func (c *Counter) Add(n int64)        { c.value.Add(n) }
func (c *Counter) Value() int64       { return c.value.Load() }
func (c *Counter) Name() string       { return c.name }
func (c *Counter) Reset()             { c.value.Store(0) }

// Gauge is a value that can go up and down.
type Gauge struct {
	mu    sync.Mutex
	value float64
	name  string
}

func NewGauge(name string) *Gauge { return &Gauge{name: name} }
func (g *Gauge) Set(v float64)    { g.mu.Lock(); g.value = v; g.mu.Unlock() }
func (g *Gauge) Inc()             { g.mu.Lock(); g.value++; g.mu.Unlock() }
func (g *Gauge) Dec()             { g.mu.Lock(); g.value--; g.mu.Unlock() }
func (g *Gauge) Add(v float64)    { g.mu.Lock(); g.value += v; g.mu.Unlock() }
func (g *Gauge) Value() float64   { g.mu.Lock(); defer g.mu.Unlock(); return g.value }
func (g *Gauge) Name() string     { return g.name }

// Histogram collects observations into buckets.
type Histogram struct {
	mu      sync.Mutex
	name    string
	buckets []float64
	counts  []uint64
	sum     float64
	count   uint64
	min     float64
	max     float64
}

func NewHistogram(name string, buckets []float64) *Histogram {
	sort.Float64s(buckets)
	return &Histogram{name: name, buckets: buckets, counts: make([]uint64, len(buckets)+1), min: math.MaxFloat64, max: -math.MaxFloat64}
}

func (h *Histogram) Observe(v float64) {
	h.mu.Lock()
	defer h.mu.Unlock()
	h.sum += v
	h.count++
	if v < h.min {
		h.min = v
	}
	if v > h.max {
		h.max = v
	}
	idx := sort.SearchFloat64s(h.buckets, v)
	h.counts[idx]++
}

func (h *Histogram) Count() uint64 { h.mu.Lock(); defer h.mu.Unlock(); return h.count }
func (h *Histogram) Sum() float64  { h.mu.Lock(); defer h.mu.Unlock(); return h.sum }
func (h *Histogram) Min() float64  { h.mu.Lock(); defer h.mu.Unlock(); return h.min }
func (h *Histogram) Max() float64  { h.mu.Lock(); defer h.mu.Unlock(); return h.max }
func (h *Histogram) Name() string  { return h.name }
func (h *Histogram) Mean() float64 {
	h.mu.Lock()
	defer h.mu.Unlock()
	if h.count == 0 {
		return 0
	}
	return h.sum / float64(h.count)
}

// Snapshot returns bucket boundaries and their counts.
func (h *Histogram) Snapshot() ([]float64, []uint64) {
	h.mu.Lock()
	defer h.mu.Unlock()
	b := make([]float64, len(h.buckets))
	c := make([]uint64, len(h.counts))
	copy(b, h.buckets)
	copy(c, h.counts)
	return b, c
}

// Registry collects named metrics.
type Registry struct {
	mu         sync.RWMutex
	counters   map[string]*Counter
	gauges     map[string]*Gauge
	histograms map[string]*Histogram
}

func NewRegistry() *Registry {
	return &Registry{counters: make(map[string]*Counter), gauges: make(map[string]*Gauge), histograms: make(map[string]*Histogram)}
}

func (r *Registry) Counter(name string) *Counter {
	r.mu.Lock()
	defer r.mu.Unlock()
	if c, ok := r.counters[name]; ok {
		return c
	}
	c := NewCounter(name)
	r.counters[name] = c
	return c
}

func (r *Registry) Gauge(name string) *Gauge {
	r.mu.Lock()
	defer r.mu.Unlock()
	if g, ok := r.gauges[name]; ok {
		return g
	}
	g := NewGauge(name)
	r.gauges[name] = g
	return g
}

func (r *Registry) Histogram(name string, buckets []float64) *Histogram {
	r.mu.Lock()
	defer r.mu.Unlock()
	if h, ok := r.histograms[name]; ok {
		return h
	}
	h := NewHistogram(name, buckets)
	r.histograms[name] = h
	return h
}

// Snapshot exports all metrics as a flat map.
func (r *Registry) Snapshot() map[string]interface{} {
	r.mu.RLock()
	defer r.mu.RUnlock()
	snap := make(map[string]interface{})
	for n, c := range r.counters {
		snap["counter."+n] = c.Value()
	}
	for n, g := range r.gauges {
		snap["gauge."+n] = g.Value()
	}
	for n, h := range r.histograms {
		snap["histogram."+n+".count"] = h.Count()
		snap["histogram."+n+".mean"] = h.Mean()
	}
	return snap
}

// CounterNames returns sorted counter names.
func (r *Registry) CounterNames() []string {
	r.mu.RLock()
	defer r.mu.RUnlock()
	names := make([]string, 0, len(r.counters))
	for n := range r.counters {
		names = append(names, n)
	}
	sort.Strings(names)
	return names
}
